### ISSUE TITLE
[Proposal] NameOf(Me)

### ISSUE THREAD

**Issue by AdamSpeight2008** [dotnet/vblang#20]
[Ported from Roslyn Repo ](https://github.com/dotnet/roslyn/issues/16130)

`NameOf(Me)` should evaluate to the Class / Structure / Module name, it is used in.

```VB
Module Exts
  Public Function Foo() As Object
    Debug.WriteLine(NameOf(Me)) ' output "Exs"
 End Function
End Module
```
*Advantage*
Doesn't require reflection. 
Doesn't require repetition of the TypeName

**Comment by jnm2** [comment]
Unless `Me.Foo()` is valid, `NameOf(Me)` would be inconsistent.

**Comment by jrmoreno1** [comment]
There is (or was) a more general request that would cover the current method/property.  Solving it for just the class/module is just too limiting.

**Comment by AnthonyDGreen** [comment]
Why is repeating the type name an issue here?

**Comment by KlausLoeffelmann** [comment]
I have a couple of questions.

What happens, if NameOf(Me) is used in a Method in Foo, and called somewhere from Bar, and that Method gets inlined? Reflection would return something different than NameOf, right? I would not like that.

And would you expect that NameOf(Me) is returning different results in derived classes? So, if you have

Class Foo
    Function FooMethod as String
          Return NameOf(Me)
    End Fuction 
End Class

Class Bar
   Inherits Foo
End Class

What happens if you call Foo.FooMethod in contrast to Bar.FooMethod? Remember: the Return-Value is prepared already at Compile-Time!

**Comment by jrmoreno1** [comment]
@AnthonyDGreen: I read "Doesn't require repetition of the TypeName" to mean "doesn't require magic strings".

**Comment by AnthonyDGreen** [comment]
@jrmoreno1 you don't need magic strings, you can just say NameOf(TypeName)

**Comment by jrmoreno1** [comment]
@AnthonyDGreen : yep, you're right. Can't remember what I was thinking, but what I said is obviously wrong.
